[SPARK-4553] [SPARK-5767] [SQL] Wires Parquet data source with the newly introduced write support for data source API

liancheng opened the pull request:
This PR migrates the Parquet data source to the new data source write support API.  Now users can also overwriting and appending to existing tables. Notice that inserting into partitioned tables is not supported yet.

When Parquet data source is enabled, insertion to Hive Metastore Parquet tables is also fullfilled by the Parquet data source. This is done by the newly introduced `HiveMetastoreCatalog.ParquetConversions` rule, which is a "proper" implementation of the original hacky `HiveStrategies.ParquetConversion`. The latter is still preserved, and can be removed together with the old Parquet support in the future.

TODO:
- [x] Update outdated comments in `newParquet.scala`.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/apache/spark/4563)

<!-- Reviewable:end -->

liancheng committed:
Passes Hive Metastore partitioning information to ParquetRelation2

Rewires Parquet data source and the new data source write support

Temporary solution for moving Parquet conversion to analysis phase

Although it works, it's so ugly... I duplicated the whole Analyzer
in Hive Context. Have to fix this.

Cleaner solution for Metastore Parquet table conversion

Fixes compilation errors introduced during rebasing

Minor cleanups

Addresses @yhuai's comments

SparkQA commented:
[Test build #27345 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27345/consoleFull) for   PR 4563 at commit [`ae17ea8`](https://github.com/apache/spark/commit/ae17ea8a3b1123845cd0be19cd1073bbaf9d0d7b).
 * This patch merges cleanly.

liancheng commented:
cc @marmbrus @rxin @yhuai

liancheng commented on sql/core/src/main/scala/org/apache/spark/sql/parquet/newParquet.scala:
```diff
@@ -89,7 +89,7 @@ class DefaultSource
       mode match {
         case SaveMode.Append =>
-          sys.error(s"Append mode is not supported by ${this.getClass.getCanonicalName}")
+          true
```
Enabling append.

liancheng commented on sql/core/src/main/scala/org/apache/spark/sql/parquet/newParquet.scala:
```diff
@@ -106,12 +106,12 @@ class DefaultSource
         path,
         data.schema.toAttributes,
-        false,
+        true,
```
Don't throw if the destination path already exists.

yhuai replied:
Use named parameter for allowExisting. 

Also, seems we will overwrite the metadata file. In the case of Append, do we still want to overwrite the metadata?

liancheng replied:
Yeah, good catch, we shouldn't overwrite the original metadata here, thanks!

liancheng commented on sql/core/src/main/scala/org/apache/spark/sql/parquet/newParquet.scala:
```diff
@@ -123,9 +123,9 @@ class DefaultSource
+private[sql] case class Partition(values: Row, path: String)
 
-private[parquet] case class PartitionSpec(partitionColumns: StructType, partitions: Seq[Partition])
+private[sql] case class PartitionSpec(partitionColumns: StructType, partitions: Seq[Partition])
```
These two are now private to `sql` package because I need them to pass Metastore partitioning information into `ParquetRelation2` while converting Metastore Parquet tables, and that part of code resides in `sql.hive` package.

liancheng commented on sql/core/src/main/scala/org/apache/spark/sql/parquet/newParquet.scala:
```diff
@@ -151,9 +151,12 @@ private[parquet] case class PartitionSpec(partitionColumns: StructType, partitio
+    paths: Seq[String],
+    parameters: Map[String, String],
+    maybeSchema: Option[StructType] = None,
+    maybePartitionSpec: Option[PartitionSpec] = None)(
```
Right now only used when converting Hive Metastore Parquet tables. If this is defined, we don't do partition discovery below.

liancheng commented on sql/core/src/main/scala/org/apache/spark/sql/parquet/newParquet.scala:
```diff
@@ -304,13 +328,17 @@ case class ParquetRelation2
+  def partitionColumns = metadataCache.partitionSpec.partitionColumns
 
-  private def isPartitioned = partitionColumns.nonEmpty
+  def partitions = metadataCache.partitionSpec.partitions
```
These 3 were made public for testing purposes.

liancheng commented on sql/core/src/main/scala/org/apache/spark/sql/parquet/newParquet.scala:
```diff
@@ -413,18 +441,21 @@ case class ParquetRelation2
             // TODO Avoids boxing cost here!
-            row.update(partitionKeyLocations(i), partValues(i))
+            val partOrdinal = requiredPartOrdinal(i)
+            row.update(partitionKeyLocations(partOrdinal), partValues(partOrdinal))
```
This is a bug fix, shouldn't always append *all* partition keys to the result row.

liancheng commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveContext.scala:
```diff
@@ -244,6 +244,7 @@ class HiveContext(sc: SparkContext) extends SQLContext(sc) {
   override protected[sql] lazy val analyzer =
     new Analyzer(catalog, functionRegistry, caseSensitive = false) {
       override val extendedRules =
+        catalog.ParquetConversions ::
```
Moved Metastore Parquet table conversion to analysis phase.

marmbrus replied:
So much cleaner....

liancheng commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala:
```diff
@@ -183,21 +177,47 @@ private[hive] class HiveMetastoreCatalog(hive: HiveContext) extends Catalog with
-        LogicalRelation(ParquetRelation2(
-          paths, Map(ParquetRelation2.METASTORE_SCHEMA -> metastoreSchema.json))(hive))
+        // convertToParquetRelation(relation)
+        relation
```
Oops, should remove the whole `if` here.

liancheng commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala:
```diff
@@ -380,13 +400,48 @@ private[hive] class HiveMetastoreCatalog(hive: HiveContext) extends Catalog with
     }
   }
 
+  object ParquetConversions extends Rule[LogicalPlan] {
```
Forgot to write ScalaDoc for this rule.

SparkQA commented:
[Test build #27346 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27346/consoleFull) for   PR 4563 at commit [`efcc8d2`](https://github.com/apache/spark/commit/efcc8d2d0c22a0675e479068c41036256ba73d23).
 * This patch merges cleanly.

SparkQA commented:
[Test build #27345 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27345/consoleFull) for   PR 4563 at commit [`ae17ea8`](https://github.com/apache/spark/commit/ae17ea8a3b1123845cd0be19cd1073bbaf9d0d7b).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class ParquetRelation2(`

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27345/
Test PASSed.

SparkQA commented:
[Test build #27346 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27346/consoleFull) for   PR 4563 at commit [`efcc8d2`](https://github.com/apache/spark/commit/efcc8d2d0c22a0675e479068c41036256ba73d23).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class ParquetRelation2(`

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27346/
Test PASSed.

liancheng commented on sql/core/src/main/scala/org/apache/spark/sql/parquet/newParquet.scala:
```diff
@@ -106,12 +106,12 @@ class DefaultSource
       ParquetRelation.createEmpty(
```
Currently we are still using some utility functions like this one from the old Parquet support code. We can move them into the new data source in the future.

liancheng commented on sql/core/src/main/scala/org/apache/spark/sql/parquet/newParquet.scala:
```diff
@@ -226,13 +247,14 @@ case class ParquetRelation2
-      partitionSpec = {
-        val partitionDirs = dataStatuses
+      partitionSpec = maybePartitionSpec.getOrElse {
+        val partitionDirs = leaves
```
Using all files rather than Parquet part-files only because there can be empty partitions. (Should we include entirely empty directories here? Namely directories don't even contain `_metadata` or `_common_metadata`.)

yhuai replied:
Is it possible that we will miss partitioning columns? If not, I think we can ignore entirely empty directories.

liancheng replied:
No, number of partition columns must be the same among all partition directories.

liancheng commented on sql/core/src/main/scala/org/apache/spark/sql/parquet/newParquet.scala:
```diff
@@ -545,13 +578,16 @@ object ParquetRelation2 {
 
-  // Default partition name to use when the partition column value is null or empty string
+  // Schema of partition keys for partitioned Hive Metastore Parquet tables.
+  val METASTORE_PARTITION_KEYS_SCHEMA = "metastorePartitionKeysSchema"
```
Ah, forgot to remove this. It's not used any more since we construct `PartitionSpec` manually and pass it to `ParquetRelation2` when converting Metastore Parquet tables.

yhuai commented on sql/core/src/main/scala/org/apache/spark/sql/parquet/newParquet.scala:
```diff
@@ -106,12 +106,12 @@ class DefaultSource
 
       val createdRelation = createRelation(sqlContext, parameters, data.schema)
-      createdRelation.asInstanceOf[ParquetRelation2].insert(data, true)
+      createdRelation.asInstanceOf[ParquetRelation2].insert(data, mode == SaveMode.Overwrite)
```
Use named parameter for overwrite.

yhuai commented on sql/core/src/main/scala/org/apache/spark/sql/parquet/newParquet.scala:
```diff
@@ -187,31 +190,49 @@ case class ParquetRelation2
         val statuses = SparkHadoopUtil.get.listLeafStatuses(fs, f.getPath).filter { f =>
           isSummaryFile(f.getPath) ||
             !(f.getPath.getName.startsWith("_") || f.getPath.getName.startsWith("."))
         }
```
Add a comment at here to explain what we are doing.

yhuai commented on sql/core/src/main/scala/org/apache/spark/sql/parquet/newParquet.scala:
```diff
@@ -187,31 +190,49 @@ case class ParquetRelation2
+      }.toArray
+      assert(baseStatuses.forall(!_.isDir) || baseStatuses.forall(_.isDir))
 
       val leaves = baseStatuses.flatMap { f =>
```
Maybe `leafSummaryFiles?`

yhuai commented on sql/core/src/main/scala/org/apache/spark/sql/parquet/newParquet.scala:
```diff
@@ -465,7 +498,7 @@ case class ParquetRelation2
 
     val job = new Job(sqlContext.sparkContext.hadoopConfiguration)
-    val writeSupport = if (schema.map(_.dataType).forall(_.isPrimitive)) {
+    val writeSupport = if (parquetSchema.map(_.dataType).forall(_.isPrimitive)) {
```
What will happen if we do `save` with `SaveMode.Overwrite` and the given `DataFrame` has a different schema with the existing one?

Should we use the schema of `data`?

yhuai replied:
Right now, we have a schema check for `SaveMode.Append` in `CreateMetastoreDataSourceAsSelect`.

liancheng replied:
For `SaveMode.Overwrite`, we are always writing to an empty directory, so there's no "existing one".

liancheng replied:
But I should use `data.schema` rather than `parquetSchema` here...

yhuai commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala:
```diff
@@ -381,12 +391,51 @@ private[hive] class HiveMetastoreCatalog(hive: HiveContext) extends Catalog with
+              relation.tableDesc.getSerdeClassName.toLowerCase.contains("parquet") =>
+          relation
+
+        case p @ PhysicalOperation(_, _, relation: MetastoreRelation)
```
Add a comment to explain when this one will be triggered?

marmbrus commented on sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetTest.scala:
```diff
@@ -23,8 +23,9 @@ import scala.reflect.ClassTag
 
-import org.apache.spark.sql.{DataFrame, SQLContext}
 import org.apache.spark.sql.catalyst.util
+import org.apache.spark.sql.sources.SaveMode
```
Is this up to date?  I thought we moved `SaveMode` to `sql.`

liancheng replied:
Yeah, rebased.

liancheng commented:
Squashed all commits to ease rebasing.

SparkQA commented:
[Test build #27515 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27515/consoleFull) for   PR 4563 at commit [`a83d290`](https://github.com/apache/spark/commit/a83d29064155ea48442ed4a2f378f00d52fa75f5).
 * This patch merges cleanly.

SparkQA commented:
[Test build #27515 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27515/consoleFull) for   PR 4563 at commit [`a83d290`](https://github.com/apache/spark/commit/a83d29064155ea48442ed4a2f378f00d52fa75f5).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class ParquetRelation2(`

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27515/
Test FAILed.

liancheng committed:
Fixes compilation error introduced during rebasing

SparkQA commented:
[Test build #27517 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27517/consoleFull) for   PR 4563 at commit [`2476e82`](https://github.com/apache/spark/commit/2476e82132512e2aa8b601619a7a94fe8967806c).
 * This patch merges cleanly.

SparkQA commented:
[Test build #27517 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27517/consoleFull) for   PR 4563 at commit [`2476e82`](https://github.com/apache/spark/commit/2476e82132512e2aa8b601619a7a94fe8967806c).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class ParquetRelation2(`

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27517/
Test PASSed.

liancheng commented:
@yhuai @marmbrus All comments have been addressed, have to squash all commits to ease rebasing. Should be good to go.

liancheng committed:
Fixes test cases which should disable off Parquet data source

SparkQA commented:
[Test build #27544 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27544/consoleFull) for   PR 4563 at commit [`fa98d27`](https://github.com/apache/spark/commit/fa98d277e5294af4b8a6460d5f02379c33da714d).
 * This patch merges cleanly.

liancheng commented:
Many thanks to @yhuai, who helped getting rid of a bug in those Parquet test suites which disable Parquet data source and fall back to the old implementation!

liancheng commented:
Removed the `withSQLConf` trick in test suites like `ParquetQuerySuite`. A simplified version of the problem can be shown as:

```scala
withSQLConf(SQLConf.PARQUET_USE_DATA_SOURCE_API -> "false") {
  test("some test") {
     ...
  }
}
```

The execution order of this snippet is:

1. `SQLConf.PARQUET_USE_DATA_SOURCE_API` is set to `false`
2. A ScalaTest test case "some test" is created
3. `SQLConf.PARQUET_USE_DATA_SOURCE_API` is reverted (removed)
4. ScalaTest starts executing test case "some test"
5. "some test" executes with the default value of `SQLConf.PARQUET_USE_DATA_SOURCE_API` configuration, which is `true`

In the last commit, I removed the `withSQLConf` trick and fall back to `beforeAll`/`afterAll`. This introduced hundreds of lines of indentation changes in several test suites, thus made this PR twice as large.

liancheng commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveStrategies.scala:
```diff
@@ -139,15 +139,19 @@ private[hive] trait HiveStrategies {
+              PhysicalRDD(plan.output, sparkContext.emptyRDD[Row]) :: Nil
+            } else {
+              hiveContext
+                .parquetFile(partitionLocations.head, partitionLocations.tail: _*)
```
This is a bug fix. When no partition is selected, `partitionLocation.head` throws. In Spark 1.2, `parquetFile` accepts a single `path` argument. In this case, `parquetFile` throws an `IllegalArgumentException` since `path` is empty. This exception is then explicitly caught below.

SparkQA commented:
[Test build #27544 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27544/consoleFull) for   PR 4563 at commit [`fa98d27`](https://github.com/apache/spark/commit/fa98d277e5294af4b8a6460d5f02379c33da714d).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class ParquetRelation2(`

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27544/
Test PASSed.

liancheng commented:
@yhuai @marmbrus Thanks for the review! I'm gonna merge this as it contains a bunch of fixes and enables full power of the new Parquet data source :)
